### PR TITLE
add test for nic binding

### PR DESCRIPTION
**Pull request by gouyang**
Signed-off-by: Guohua Ouyang <gouyang@redhat.com>

### CODE REVIEW

File: frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    await $(vmView.networkBindingId).click();
+
+    await selectDropdownOption(vmView.networkTypeDropdownId, networkInterface.networkDefinition);
+
+    await $(vmView.networkBindingId).click();
+    options = await getDropdownOptions(vmView.networkBindingId);
+    expect(options).toEqual([networkBindingMethod.default, networkBindingMethod.sriov]);
+    await $(vmView.networkBindingId).click();
+
+    await selectDropdownOption(vmView.networkBindingId, networkBindingMethod.default);
```
Comment (rhrazdil) [inline frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts]: Again, can we use `wizard.addNic` ?

File: frontend/integration-tests/views/kubevirt/wizard.view.ts [hunk 2]
```diff
@@ -77,3 +77,8 @@ export const cloneVmWarning = cloneVmDialog.$('.alert-warning');
 
 // VMs List view
 export const firstRowVMStatus = $('div.co-m-row:first-child > div:first-child > div:nth-child(3)');
+
+// VM wizard network
+export const dropdownOptionID = (tableType, rowNumber) => `#${tableType}-edit-${rowNumber}-row`;
+export const dropdownOption = (tableType, rowNumber) => $(dropdownOptionID(tableType, rowNumber));
+export const editableTableCell = $('.kubevirt-editable-table').$$('td');
```
Comment (rhrazdil) [inline frontend/integration-tests/views/kubevirt/wizard.view.ts]: editableTableCell is maybe not necessary if `activateTableRow` is used.

File: frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+    await browser.get(`${appHost}/k8s/ns/${testName}/virtualmachines`);
+    await isLoaded();
+    await wizard.openWizard();
+
+    // Basic Settings for VM
+    await wizard.fillName(vmName);
+    await wizard.fillDescription(testName);
+    await wizard.selectProvisionSource(provisionSource);
+    //await wizard.selectOperatingSystem(basicVmConfig.operatingSystem);
```
[the remark this answers is not in the document]
Comment (gouyang) [reply frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts]: Just forget to uncomment this line.

File: frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    await wizard.selectOperatingSystem('Fedora 29');
```
[the remark this answers is not in the document]
Comment (gouyang) [reply frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts]: Will remove this line and uncomment above line.

File: frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    await wizard.selectWorkloadProfile(basicVmConfig.workloadProfile);
+    await wizard.next();
+
+    // Networking
+    let rowsCount = await wizardView.tableRowsCount();
+    let bindingID = wizardView.dropdownOptionID('binding', rowsCount);
+    // Pod networking default
+    expect(wizardView.editableTableCell.get(3).getText()).toEqual(networkBindingMethod.pod);
+    await wizardView.editableTableCell.get(3).click();
+    await wizardView.editableTableCell.get(3).click();
```
[the remark this answers is not in the document]
Comment (gouyang) [reply frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts]: This is not to activate the tableRow, it's clicking twice to get the dropdown menu.

File: frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    await wizardView.selectTableDropdownAttribute(rowsCount, 'network', networkInterface.networkDefinition);
+    expect(wizardView.dropdownOption('binding', rowsCount).getText()).toEqual(networkBindingMethod.default);
+    await $(bindingID).click();
+    options = await getDropdownOptions(bindingID);
+    expect(options).toEqual([networkBindingMethod.default, networkBindingMethod.sriov]);
+    await $(bindingID).click();
+
+    await wizardView.selectTableDropdownAttribute(rowsCount, 'binding', networkBindingMethod.default);
+    await wizardView.setTableInputAttribute(rowsCount, 'mac', networkInterface.mac);
+    await wizardView.setTableInputAttribute(rowsCount, 'name', networkInterface.name);
```
[the remark this answers is not in the document]
Comment (gouyang) [reply frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts]: It has some verifications in the process of adding nic, so `wizard.addNic` is not suitable.

File: frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    // Create VM
+    await wizard.next();
+    leakedResources.add(JSON.stringify({name: vmName, namespace: testName, kind: 'vm'}));
+  });
+
+  it('Check nic binding method on vm nic tab', async() => {
+    await vm.navigateToTab(vmView.nicsTab);
+    await isLoaded();
+
+    expect(resourceRows.first().$('div:nth-child(4)').getText()).toEqual(networkBindingMethod.pod);
```
[the remark this answers is not in the document]
Comment (gouyang) [reply frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts]: How to name `$('div:nth-child(4)')` if put it in `wizard.view.ts`?

File: frontend/integration-tests/views/kubevirt/wizard.view.ts (hunk 2 above)
Comment (gouyang) [reply frontend/integration-tests/views/kubevirt/wizard.view.ts]: `activateTableRow` does not return an element, I need an element to do do some assertion in #L52.

File: frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts (hunk 1 above)
Comment (gouyang) [reply frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts]: `wizard.addNic` is not suitable because it has some verifications in the process.

File: frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    expect(options).toEqual([networkBindingMethod.pod, networkBindingMethod.default, networkBindingMethod.sriov]);
+    await wizardView.cancelButton.click();
+
+    // Non pod networking
+    await wizardView.createNIC.click();
+    rowsCount = await wizardView.tableRowsCount();
+    bindingID = wizardView.dropdownOptionID('binding', rowsCount);
+    await $(bindingID).click();
+    options = await getDropdownOptions(bindingID);
+    expect(options).toEqual([networkBindingMethod.pod, networkBindingMethod.default, networkBindingMethod.sriov]);
```
[the remark this answers is not in the document]
Comment (gouyang) [reply frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts]: Updated.

File: frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+    expect(resourceRows.first().$('div:nth-child(4)').getText()).toEqual(networkBindingMethod.pod);
+    expect(resourceRows.get(1).$('div:nth-child(4)').getText()).toEqual(networkBindingMethod.default);
+
+    await vm.removeNic(networkInterface.name);
+    await isLoaded();
+
+    await vmView.createNic.click();
+    await $(vmView.networkBindingId).click();
+    let options = await getDropdownOptions(vmView.networkBindingId);
+    expect(options).toEqual([networkBindingMethod.pod, networkBindingMethod.default, networkBindingMethod.sriov]);
```
[the remark this answers is not in the document]
Comment (gouyang) [reply frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts]: updated

File: frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (rhrazdil) [reply frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts]: Would something like this work? 
```
...
// Networking tab
export const createNIC = $('#create-network-btn');
export const pxeNICDropdownId = '#pxe-nic-dropdown';

const tableRows = $('.kubevirt-editable-table > tbody').$$('tr');
export const nicName = (row) => tableRows.get(row-1).$('div:nth-child(1)');
export const nicMac = (row) => tableRows.get(row-1).$('div:nth-child(2)');
export const nicNetwork = (row) => tableRows.get(row-1).$('div:nth-child(3)');
export const nicBinding = (row) => tableRows.get(row-1).$('div:nth-child(4)');
...

File: frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+    await wizard.selectOperatingSystem(basicVmConfig.operatingSystem);
+    await wizard.selectFlavor(basicVmConfig.flavor);
+    await wizard.selectWorkloadProfile(basicVmConfig.workloadProfile);
+    await wizard.next();
+
+    // Networking
+    // Pod networking
+    let rowsCount = await wizardView.tableRowsCount();
+    let bindingID = wizardView.dropdownOptionID('binding', rowsCount);
+    const podNetworkBinding = wizardView.itemInTable(0, networkWizardTabCol.binding - 1);
```
Comment (suomiy) [inline frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts]: why not start counting from zero in `networkWizardTabCol`, etc. ?

File: frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+    // Pod networking default binding is masquerade
+    expect(podNetworkBinding.getText()).toEqual(networkBindingMethod.pod);
+    await podNetworkBinding.click();
```
Comment (suomiy) [inline frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts]: maybe add this comment ? `// open a row`

File: frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+    await wizardView.cancelButton.click();
+
+    // Non pod networking
+    await wizardView.createNIC.click();
+    rowsCount = await wizardView.tableRowsCount();
+    bindingID = wizardView.dropdownOptionID('binding', rowsCount);
+    await $(bindingID).click();
+    options = await getDropdownOptions(bindingID);
+    // Without network selected, binding method masquerade, bridge and sriov are available
+    assertUnorderListEqual(options, podNetworkList);
```
Comment (suomiy) [inline frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts]: maybe add a differently named variable? allNetworkList ?

File: frontend/integration-tests/tests/kubevirt/utils.ts [hunk 13]
```diff
@@ -170,3 +173,10 @@ export const waitForStringInElement = (element: ElementFinder, needle: string) =
     return content.includes(needle);
   };
 };
+
+export function assertUnorderListEqual(elementList, targetList) {
+  expect(elementList.length).toEqual(targetList.length);
```
[the remark this answers is not in the document]
Comment (gouyang) [reply frontend/integration-tests/tests/kubevirt/utils.ts]: Good point, while writing the function I'm aware of it's not necessary to do it at all. I removed this function and sort the list in the test file.

File: frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+describe('Test VM network interface binding method', () => {
+  const leakedResources = new Set<string>();
+  const wizard = new Wizard();
+  const vmName = `vm-${testName}`;
+  const vm = new VirtualMachine(vmName, testName);
+  const podNetworkList = [networkBindingMethod.pod, networkBindingMethod.default, networkBindingMethod.sriov];
+  const networkList = [networkBindingMethod.default, networkBindingMethod.sriov];
+
+  beforeAll(async() => {
+    execCommandFromCli(`echo '${JSON.stringify(testNad)}' | kubectl create -f -`);
```
[the remark this answers is not in the document]
Comment (gouyang) [reply frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts]: there are functions `createResources` and `deleteResources`, but the `deleteResources` cannot delete a vm which isn't derived from `getVmManifest`.

File: frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts (hunk 10 above)
Comment (gouyang) [reply frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts]: updated.

File: frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts (hunk 11 above)
Comment (gouyang) [reply frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts]: added

File: frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts (hunk 12 above)
Comment (gouyang) [reply frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts]: updated

File: frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (gouyang) [reply frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts]: updated

File: frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (gouyang) [reply frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts]: updated

File: frontend/integration-tests/tests/kubevirt/mocks.ts [hunk 15]
```diff
@@ -266,6 +267,40 @@ export const multusNetworkInterface = {
   networkDefinition: multusNad.metadata.name,
 };
 
+export const networkBindingMethod = {
+  pod: 'masquerade',
```
Comment (rhrazdil) [inline frontend/integration-tests/tests/kubevirt/mocks.ts]: I think it would be more clear to name the key by their value

File: frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (rhrazdil) [reply frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts]: ```deleteResources``` deletes anything that has `name, namespace and kind` properties.
I'm a bit confused what is the point of execCommandFromCli method because it doesn't provide higher abstraction than execSync does, moreover it consumes the potential exception so if something happens, the test execution goes on as if nothing happened. I think we should get rid of it.

File: frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts (hunk 11 above)
Comment (suomiy) [reply frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts]: I meant to add it so it explains why we click on podNetworkBinding twice. Not sure if it is apparent from current comment

File: frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (suomiy) [reply frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts]: good to know that these utility functions were created. Maybe we could use them here?

\+ having methods for single resource would be beneficial as well

File: frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts (hunk 11 above)
Comment (gouyang) [reply frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts]: 1st click enable the field, 2nd click get the dropdown list. Update the comment.

File: frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (gouyang) [reply frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts]: Updated `deleteResources` and use it now.

File: frontend/integration-tests/tests/kubevirt/mocks.ts (hunk 15 above)
Comment (gouyang) [reply frontend/integration-tests/tests/kubevirt/mocks.ts]: I updated it by using the same string for the key and the value.

File: frontend/integration-tests/tests/kubevirt/utils.ts [hunk 16]
```diff
@@ -85,6 +97,9 @@ export async function fillInput(elem: ElementFinder, value: string) {
       throw Error(`Failed to fill input with value: '${value}'.`);
     }
     await browser.wait(until.elementToBeClickable(elem));
+    // ensure the element is enabled, otherwise it fails to send key.
+    // https://bugs.chromium.org/p/chromedriver/issues/detail?id=1963
+    await elem.click();
```
Comment (suomiy) [inline frontend/integration-tests/tests/kubevirt/utils.ts]: how does clicking on an element enable it?

File: frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts (hunk 11 above)
Comment (suomiy) [reply frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts]: 1st click activates the row

File: frontend/integration-tests/tests/kubevirt/utils.ts [hunk 17]
```diff
@@ -56,12 +56,24 @@ export function createResources(resources) {
 export function deleteResources(resources) {
   resources.forEach(resource => {
     const kind = resource.kind === 'NetworkAttachmentDefinition' ? 'net-attach-def' : resource.kind;
-    execSync(`kubectl delete -n ${resource.metadata.namespace} --cascade ${kind} ${resource.metadata.name}`);
+    let namespace = '';
```
[the remark this answers is not in the document]
Comment (rhrazdil) [reply frontend/integration-tests/tests/kubevirt/utils.ts]: @suomiy @gouyang  isn't there a better way to solve this? These delete/create method were intended for manifest objects, which have of course structure `kind, metadata.name, metadata.namespace`, not for class instances, at least with their current implementation.
- Isn't a better solution to provide our classes with metadata property that would contain name and namespace? Instead of adding complicated logic?
- Or maybe we distinguish between resource manifest and class instance so that if it's a class instance we would call `instance.delete()` class method that would handle the deletion.

[the remark this answers is not in the document]
Comment (suomiy) [reply frontend/integration-tests/tests/kubevirt/utils.ts]: I guess you are right. It is better to have united interface.

I propose to add this method to `DetailView`:

```js
  asResource() {
    return {
      kind: this.kind,
      metadata: {
        namespace: this.namespace,
        name :this.name,
      },
    };
  }
```

File: frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts (hunk 11 above)
Comment (gouyang) [reply frontend/integration-tests/tests/kubevirt/vm.nic.binding.scenario.ts]: done

File: frontend/integration-tests/tests/kubevirt/utils.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (gouyang) [reply frontend/integration-tests/tests/kubevirt/utils.ts]: Even we have another method for the class to delete the vm instance, it need separate lines of code to delete the resources like testNad, vm. How about just use `execSync`, it's quite directly and simple to read.  

Let's get it out of the scope of this PR, I added `vm.action('Delete')` instead of deleting the vm in `deleteResources`.

File: frontend/integration-tests/tests/kubevirt/utils.ts (hunk 16 above)
Comment (gouyang) [reply frontend/integration-tests/tests/kubevirt/utils.ts]: @suomiy Does it not work? I have tested it and it works well.
@rhrazdil This is how I solved `Element Not interactable` issue, if you have a better way then I will remove this.

Comment (suomiy) [reply frontend/integration-tests/tests/kubevirt/utils.ts]: I am more interested about how it works. Why is the referenced chromium bug relevant?

File: frontend/integration-tests/tests/kubevirt/utils.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (suomiy) [reply frontend/integration-tests/tests/kubevirt/utils.ts]: > Even we have another method for the class to delete the vm instance, it need separate lines of code to delete the resources like testNad, vm. How about just use execSync, it's quite directly and simple to read.

it is prone to mistakes and having an abstraction for simple things like CRUD is easier to read IMO.

I think, it is not necessary to delete VMs in UI when testing nics.

@rhrazdil your opinion?

File: frontend/integration-tests/tests/kubevirt/utils.ts (hunk 16 above)
Comment (rhrazdil) [reply frontend/integration-tests/tests/kubevirt/utils.ts]: I think the problem here is not that an input element is disabled (because disabled element is not clickable, so clicking it won't enable it), but the input elements in wizard network/storage  tabs are in a strange state of `enabled and yet unfocusable` (Solely my description :) ).
The input field needs to be clicked twice, as the first click doesn't seem to do anything, that is why we get ElementNotInteractable and that is also why clicking the element first makes it work. 
So to sum it up, I don't think this is related to the chromium bug, because this will behave the same in gecko driver with firefox, and I don't have a better solution, because I don't think there is a generic way how to check that this is the malformed input without clicking it first. And since clicking all input elements should not do any harm, I think this is acceptable. Unless somebody has better solution

Comment (suomiy) [reply frontend/integration-tests/tests/kubevirt/utils.ts]: so tables again :)

> And since clicking all input elements should not do any harm, I think this is acceptable.

yup

This should be fixed once we utilize pf4 tables so please add a TODO here and please remove the referenced bug.

File: frontend/integration-tests/tests/kubevirt/utils.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (rhrazdil) [reply frontend/integration-tests/tests/kubevirt/utils.ts]: @gouyang @suomiy I tend to agree, this is a small change that will allow removing created resources in a uniformed fashion. Also, this way we'll reduce execution time without losing coverage.

File: frontend/integration-tests/tests/kubevirt/utils.ts (hunk 16 above)
Comment (gouyang) [reply frontend/integration-tests/tests/kubevirt/utils.ts]: done.

File: frontend/integration-tests/tests/kubevirt/utils.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (gouyang) [reply frontend/integration-tests/tests/kubevirt/utils.ts]: Done by add `asResource` into `DetailView`.

**Comment by suomiy** [comment]
please rebase

**Comment by gouyang** [comment]
@suomiy rebase is done.
